schemas/telemetry/untrusted-modules/untrusted-modules.4.parquetmr.txt

```diff
@@ -0,0 +1,313 @@
+      optional boolean isDefaultBrowser;
+      optional boolean addonCompatibilityCheckEnabled;
+      optional binary locale (UTF8);
+      optional boolean e10sEnabled;
+      optional boolean blocklistEnabled;
+      optional binary defaultSearchEngine (UTF8);
+      optional boolean telemetryEnabled;
+      optional binary e10sCohort (UTF8);
+      optional binary searchCohort (UTF8);
+      optional int64 e10sMultiProcesses;
+      optional group defaultSearchEngineData {
+        optional binary name (UTF8);
+        optional binary loadPath (UTF8);
+        optional binary submissionURL (UTF8);
+        optional binary origin (UTF8);
+      }
+      optional group update {
+        optional boolean autoDownload;
+        optional binary channel (UTF8);
+        optional boolean enabled;
+      }
+      optional group sandbox {
+        optional int64 effectiveContentProcessLevel;
+      }
+    }
+    required group system {
+      optional boolean isWow64;
+      optional int64 memoryMB;
+      optional int64 virtualMaxMB (UINT_64);
+      optional binary appleModelId (UTF8);
+      optional group os {
+        optional binary name (UTF8);
+        optional binary locale (UTF8);
+        optional int64 windowsUBR;
+        optional int64 installYear;
+        optional int64 servicePackMajor;
+        optional int64 windowsBuildNumber;
+        optional int64 servicePackMinor;
+      }
+    }
+    optional group gfx {
+      optional group features {
+        optional group advancedLayers {
+          optional binary status (UTF8);
+        }
+        optional group webrender {
+          optional binary status (UTF8);
+        }
+        optional group wrQualified {
+          optional binary status (UTF8);
+        }
+        optional binary compositor (UTF8);
+        optional group d3d11 {
+          optional binary status (UTF8);
+          optional boolean textureSharing;
+          optional int64 version;
+          optional boolean blacklisted;
+          optional boolean warp;
+        }
+        optional group gpuProcess {
+          optional binary status (UTF8);
+        }
+        optional group d2d {
+          optional binary status (UTF8);
+          optional binary version (UTF8);
+        }
+      }
+      optional boolean DWriteEnabled;
+      optional boolean D2DEnabled;
+      optional binary ContentBackend (UTF8);
+      optional group adapters (LIST) {
+        repeated group list {
+          required group element {
+            optional binary description (UTF8);
+            optional binary vendorID (UTF8);
+            optional binary deviceID (UTF8);
+            optional binary subsysID (UTF8);
+            optional int64 RAM (UINT_64);
+            optional binary driver (UTF8);
+            optional binary driverVersion (UTF8);
+            optional binary driverDate (UTF8);
+            optional boolean GPUActive;
+          }
+        }
+      }
+      optional group monitors (LIST) {
```

Comment:
On Linux, monitors is a single object; see [0/1]. That will cause those pings to be discarded.

[0] https://github.com/mozilla-services/mozilla-pipeline-schemas/blob/dev/templates/include/telemetry/environment.1.schema.json#L509
[1] https://firefox-source-docs.mozilla.org/toolkit/components/telemetry/telemetry/data/environment.html

Reply:
The Untrusted Modules ping is Windows-only, but I'd prefer to be correct in any case. The code that generates this structure suggests it's always an array:

https://searchfox.org/mozilla-central/rev/4e094f66ced333d69b24cd49273789e3a1173dfc/widget/nsIGfxInfo.idl#69
https://searchfox.org/mozilla-central/rev/4e094f66ced333d69b24cd49273789e3a1173dfc/widget/GfxInfoBase.cpp#1199

I wonder if maybe it was an object (not an array) in previous versions? In that case I think it should be OK as-is.

The comment about Linux I think just means even if you have multiple monitors on Linux, only 1 monitor object will be returned representing the whole virtual screen.

Thoughts?